src/acp/rule.test.ts

```diff
@@ -0,0 +1,704 @@
+): Policy => {
+  const mockPolicy = createThing({ url });
+  if (rules?.forbidden) {
+    addAll(mockPolicy, ACP_NONE, rules.forbidden);
+  }
+  if (rules?.optional) {
+    addAll(mockPolicy, ACP_ANY, rules.optional);
+  }
+  if (rules?.required) {
+    addAll(mockPolicy, ACP_ALL, rules.required);
+  }
+  return mockPolicy;
+};
+
+describe("addForbiddenRuleToPolicy", () => {
+  it("adds the rule in the forbidden rules of the policy", () => {
+    const myPolicy = addForbiddenRuleToPolicy(
+      mockPolicy("https://some.pod/policy-resource#policy"),
+      mockRule("https://some.pod/rule-resource#rule")
+    );
+    expect(
+      myPolicy.has(
+        DataFactory.quad(
+          DataFactory.namedNode("https://some.pod/policy-resource#policy"),
+          DataFactory.namedNode(ACP_NONE),
+          DataFactory.namedNode("https://some.pod/rule-resource#rule")
+        )
+      )
+    ).toBe(true);
+  });
+
+  it("does not remove the existing forbidden rules", () => {
+    const mockedPolicy = mockPolicy("https://some.pod/policy-resource#policy", {
+      forbidden: [mockRule("https://some.pod/rule-resource#another-rule")],
+    });
+    const myPolicy = addForbiddenRuleToPolicy(
+      mockedPolicy,
+      mockRule("https://some.pod/rule-resource#a-rule")
+    );
+    expect(
+      myPolicy.has(
+        DataFactory.quad(
+          DataFactory.namedNode("https://some.pod/policy-resource#policy"),
+          DataFactory.namedNode(ACP_NONE),
+          DataFactory.namedNode("https://some.pod/rule-resource#another-rule")
+        )
+      )
+    ).toBe(true);
+  });
+
+  it("does not change the existing required and optional rules", () => {
+    const mockedPolicy = mockPolicy("https://some.pod/policy-resource#policy", {
+      optional: [mockRule("https://some.pod/rule-resource#optional-rule")],
+      required: [mockRule("https://some.pod/rule-resource#required-rule")],
+    });
+    const myPolicy = addForbiddenRuleToPolicy(
+      mockedPolicy,
+      mockRule("https://some.pod/rule-resource#forbidden-rule")
+    );
+    expect(
+      myPolicy.has(
+        DataFactory.quad(
+          DataFactory.namedNode("https://some.pod/policy-resource#policy"),
+          DataFactory.namedNode(ACP_ALL),
+          DataFactory.namedNode("https://some.pod/rule-resource#required-rule")
+        )
+      )
+    ).toBe(true);
+    expect(
+      myPolicy.has(
+        DataFactory.quad(
+          DataFactory.namedNode("https://some.pod/policy-resource#policy"),
+          DataFactory.namedNode(ACP_ANY),
+          DataFactory.namedNode("https://some.pod/rule-resource#optional-rule")
+        )
+      )
+    ).toBe(true);
+  });
+
+  it("does not change the input policy", () => {
+    const myPolicy = mockPolicy("https://some.pod/policy-resource#policy");
+    addForbiddenRuleToPolicy(
+      myPolicy,
+      mockRule("https://some.pod/rule-resource#rule")
+    );
+    expect(myPolicy.size).toEqual(0);
```

Comment:
I found this test kinda confusing (i.e. was creating a policy, calling an `addXXX()` function, and then asserting on the size of something being zero - so it just **_feels_** a bit jarring). I assume it's just testing that the function-under-test doesn't mutate it's input, which I guess is fine (although it feels kinda strange - i.e. do we test every single function we write to ensure it doesn't mutate it's input?). But just to make the intent clearer, I'd rewrite it as:
```
it("does not change the input policy", () => {
    const myPolicy = mockPolicy("https://some.pod/policy-resource#policy");
    const myPolicySize = myPolicy.size;
    const newPolicy = addForbiddenRuleToPolicy(
      myPolicy,
      mockRule("https://some.pod/rule-resource#rule")
    );
    expect(myPolicy.size).toEqual(myPolicySize);
    expect(newPolicy.size).toEqual(myPolicySize + 1);
};
```
And maybe I'd rewrite the test description as:
```
it("adding rule to policy does not mutate the input policy", ...
```

Reply:
I updated it in 556fa15647a426615498bb8ea4d01f4922b4edd3. I kept checking the size of the newly created policy out, because that is already tested for in the test verifying that the function actually adds the rule to the policy, so it would make this test somewhat redundant and testing for several things.